enhancements/proxy-cache.md

```diff
@@ -0,0 +1,124 @@
+  Other goals will be implemented subsequently based on the work of this proposal.
+* Cached images are read-only.
+
+## Design Details
+
+The expected pull overview is depicted as below:
+![](https://user-images.githubusercontent.com/11522230/145866763-58f44c94-839b-4edb-a95b-b9c3648cf187.png)
+Design credits: @fmissi
+
+A user initiates a pulls of an image(say a `postgres:14` image) from an upstream repository on Quay. The repository is checked to see if the image is present.
+1. If the image does not exist, a fresh pull is initiated.
+  * The user is authenticated into the upstream registry and all the layers of `postgres:14` are pulled.
+  * The pulled layers are saved to cache and served to the user in parallel.
+  * This is depicted as below:
+    ![](https://user-images.githubusercontent.com/11522230/145871778-da01585a-7b1b-4c98-903f-809c214578da.png)
+    Design credits: @fmissi
+
+2. If the image exists in cache:
+  * A user can rely on Quay's cache to stay coherent with the upstream source so that I transparently get newer images from the cache
+     when tags have been overwritten in the upstream registry immediately or after a certain period of time.
+  * If the upstream image and cached version of the image are same:
+    * No layers are pulled from the upstream repository and the cached image is served to the user.
+
+  * If the upstream image and cached version of the image are different:
+    * The user is authenticated into the upstream registry and only the changed layers of `postgres:14` are pulled.
+    * The new layers are updated in cache and served to the user in parallel.
+  * This is depicted as below:
+    ![](https://user-images.githubusercontent.com/11522230/145872216-31350e08-6746-4e34-aebf-e59a7bf6b372.png)
+    Design credits: @fmissi
+
+3. If user initiates a pull when the upstream registry which is down:
+  * If the pull happens with the configured staleness period, the image stored in cache is served.
+  * If the pull happens after the configured staleness period, the error is propagated to the user.
+  * This is depicted as below:
+    ![](https://user-images.githubusercontent.com/11522230/145878373-c23d094b-709d-4859-b875-013ea33e34f7.png)
+    Design credits: @fmissi
+
+A quay admin can leverage quota of an organization to limit cache size so the backend storage consumption remains predictable
+by discarding images from the cache according to least recently used frequency.
+This is depicted as below:
+![](https://user-images.githubusercontent.com/11522230/145884935-df19297f-96b5-4c1c-9cdc-e199e04df176.png)
+Design credits: @sdadi
+
+A user initiates a pulls of an image(say a `postgres:14` image) from an upstream repository on Quay. If the storage consumption of the organization
+is beyond the quota limit, images in the namespace are removed based on least recently used, to make space for `postgres:14` to be cached.
```

Comment:
Does the current implementation keep track of when the image was last used?

Reply:
We log it in the audit logs. But there isn't an easy way to query when a tag was last pulled. And that seems expensive to track. Given that every pull would need to make some sort of update on the database.

Reply:
We're kind of dependant on that information though, we could also use pull frequence but I'm not sure we have that either.
@Sunandadadi I think we need to add this one as an open question as well.

Reply:
Noted, I will add it.